### PR TITLE
executor: fix the incorrect untouch used in optimistic transactions (#30447)

### LINKED ISSUE #30410: Lazy constraint check may cause data inconsistency
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
/* t */ drop table if exists t;
/* t */ create table t (c0 int, c1 varchar(20), c2 varchar(20), unique key(c0), key(c2));
/* t */ insert into t (c0, c1, c2) values (1, null, 'green');

/* t */ set tidb_constraint_check_in_place=0;
/* t */ begin optimistic;
/* t */ insert into t (c0, c1, c2) values (1, 'red', 'white');
/* t */ delete from t where c1 is null;
/* t */ update t set c2 = 'green' where c2 between 'purple' and 'white';
/* t */ commit;
/* t */ admin check table t;
```

Note that tidb_constraint_check_in_place is off by default!

### 2. What did you expect to see? (Required)

No error reported by admin check table.

### 3. What did you see instead (Required)

```sql
/* t */ admin check table t;
-- t >> E8133: handle 1, index:types.Datum{k:0x1, decimal:0x0, length:0x0, i:1, collation:"", b:[]uint8(nil), x:interface {}(nil)} != record:<nil>
```

### 4. What is your TiDB version? (Required)

master (28446605c2e019e84e934602f4bd136294a61ff3)

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #30447 to release-5.2
You can switch your code base to this Pull Request by using [git-extras](https://github.com/tj/git-extras):
```bash
# In tidb repo:
git pr https://github.com/pingcap/tidb/pull/30911
```

After apply modifications, you can push your change to this PR via:
```bash
git push git@github.com:ti-srebot/tidb.git pr/30911:release-5.2-2b7ce8e61ffb
```

---

### What problem does this PR solve?

Issue Number: close #30410

Problem Summary:

When the optimistic transaction is used, the `untouched key` optimization is not compatible with lazy existence check and results in data inconsistency.

### What is changed and how it works?

Disable the `untouched` optimization for this situation when the `presumeNotExist` flag is set.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

Documentation

### Release note

```release-note
Fix the data inconsistency caused by invalid usage of lazy existence check and untouch key optimization.
```

### CODE REVIEW

File: session/session.go [hunk 1]
```diff
@@ -23,6 +23,7 @@ import (
 	"crypto/tls"
 	"encoding/json"
 	"fmt"
+	"github.com/pingcap/tidb/store/driver/txn"
```
Comment (you06) [inline session/session.go:26]: Sort the imports.

**Review by you06 — APPROVED** [review]

**Review by youjiali1995 — APPROVED** [review]

**Pull request merged**